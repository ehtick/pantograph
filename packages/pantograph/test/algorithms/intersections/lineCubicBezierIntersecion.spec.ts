import { describe, it, expect } from "vitest";

import { lineCubicBezierIntersection } from "../../../src/algorithms/intersections/lineCubicBezierIntersection.js";
import { Line } from "../../../src/models/segments/Line.js";
import { CubicBezier } from "../../../src/models/segments/CubicBezier.js";
//import { debugImg, dpnt } from "../../debug.js";

describe("lineCubicBezierIntersection", () => {
  it("should find intersections in a simple case", () => {
    const curve = new CubicBezier([0, 0], [1, 1], [1, 0.1], [0, 1.2]);
    const line = new Line([0, -0.5], [1, 1.5]);

    const intersections = lineCubicBezierIntersection(line, curve);

    //debugImg([curve, line, ...intersections.map((i) => dpnt(i))]);

    expect(intersections.length).toBe(3);
    expect(intersections).toMatchSnapshot();
  });

  it("should find intersections on an extremity", () => {
    const curve = new CubicBezier([0, 0], [1, 1], [1, 0.1], [0, 1.2]);
    const line = new Line([0, -0.5], [0, 0.5]).rotate(45);

    const intersections = lineCubicBezierIntersection(line, curve);

    //debugImg([curve, line, ...intersections.map((i) => dpnt(i))]);

    expect(intersections.length).toBe(1);
    expect(intersections[0]).toBeVector([0, 0]);
  });

  it("should not find intersections when there are none", () => {
    const curve = new CubicBezier([0, 0], [1, 1], [1, 0.1], [0, 1.2]);
    const line = new Line([-10, -0.5], [2, 0.5]).rotate(45);

    const intersections = lineCubicBezierIntersection(line, curve);

    //debugImg([curve, line, ...intersections.map((i) => dpnt(i))]);

    expect(intersections.length).toBe(0);
  });
});
