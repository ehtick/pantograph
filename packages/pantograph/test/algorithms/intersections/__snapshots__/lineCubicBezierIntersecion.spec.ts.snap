// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineCubicBezierIntersection > should find intersections in a simple case 1`] = `
[
  [
    0.7618019930204666,
    1.023603986040933,
  ],
  [
    0.49848113110444375,
    0.49696226220888706,
  ],
  [
    0.2902834858334786,
    0.0805669716669567,
  ],
]
`;
